lib/index.js

```diff
@@ -344,14 +344,29 @@ function Postgres(a, b) {
 
     listeners[channel] = [fn]
     return query({ raw: true }, getListener(), 'listen ' + escape(channel))
+      .then((result) => {
+        result.unlisten = unlisten
+        return result
+      })
+
+    function unlisten() {
+      const index = listeners[channel].indexOf(fn)
+      if (index !== -1)
+        listeners[channel].splice(index, 1)
+      if (!listeners[channel].length) {
+        delete listeners[channel]
+        return query({ raw: true }, getListener(), 'unlisten ' + escape(channel))
+      }
+      return Promise.resolve()
+    }
   }
 
   function getListener() {
     if (listener)
       return listener
 
     listener = Connection(Object.assign({
-      onnotify: (c, x) => c in listeners && listeners[c].forEach(fn => fn(x)),
+      onnotify: (c, x) => c in listeners && listeners[c].slice(0).forEach(fn => fn(x)),
```

Comment:
What is the purpose of making a copy of the array here?

Reply:
this is standard issue with event buses. I do mutate an array in unlisten. If somebody decides to do this:

```
const s = await sql.listen(“test”, () => s.unlisten());
```

It will mutate an array during iteration, what will cause indexes shift and results in some listeners not being called. I had a similar issue in my lib: https://github.com/stalniy/casl/issues/159

Reply:
Ah, of course! That's so subtle, but makes perfect sense. Thanks!

Reply:
Why not making a copy of the array in `unlisten`, so that it is copied less often?

`listeners[channel] = listeners[channel].slice(0).splice(index, 1)`

https://github.com/porsager/postgres/pull/155/files#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R357-R358